Rewrite ReactTreeTraversal-test.js using public APIs

timjacobi opened the pull request:
Hej!

This is a super interesting test to look at since the tests cover 3 different areas when you try to rewrite them using only public APIs. So far I have converted 1/3 of the tests to use public APIs only. The rest I'd like to discuss.

### `traverseTwoPhase`
These were fairly straightforward to rewrite. I could even used the same `expectedCalls` arrays for the expectations on the mock.

The only thing that could be a bit nicer is the calls to `eventHandler`. I would like to factor out passing the arguments but #11663 prevents me from passing the event phase dynamically.

### `traverseEnterLeave`
In order to test them in the same fashion as most other _public API_ tests I would have to dispatch `onmouseenter` and `onmouseleave` events against the DOM nodes and see which event handlers React calls. Unfortunately dispatching these event types doesn't have an effect and looking at `EnterLeaveEventPlugin.js` it seems like _enter_ and _leave_ events are implemented using `onmouseover` and `onmouseout`. 
The tests seem to verify the _capturing_ and _bubbling_ behaviour which are inherently different between `onmouseenter` and `onmouseover` as well as `onmouseleave` and `onmouseout`. I would be really thank for some ideas on how we could solve this.

### `getFirstCommonAncestor` aka `getLowestCommonAncestor`
This function is only used by the `ResponderEventPlugin` which is in turn only used by React Native. This doesn't mean we can't test it only with public APIs but we might struggle with that in our _JSDOM_ environment. Again would be great to discuss alternatives :)

gaearon commented:
>Unfortunately dispatching these event types doesn't have an effect and looking at EnterLeaveEventPlugin.js it seems like enter and leave events are implemented using onmouseover and onmouseout.

This sounds right.

>I would be really thank for some ideas on how we could solve this.

I'm not sure I understand the problem. The traversal tests verify that we correctly walk the tree for different "types" of walks (enter-leave is one example). It seems right to me that you would test it by listening on React `onMouseEnter` and `onMouseLeave`, but dispatch whatever native events necessary. Maybe `EnterLeavePlugin-test` could be useful as inspiration? Or could you clarify what is the problem you ran into?

>This function is only used by the ResponderEventPlugin which is in turn only used by React Native. 

If you search for `require('react-native')` you'll find a few places where we load its mock. Maybe some of these tests could be helpful?

timjacobi commented:
> Or could you clarify what is the problem you ran into?
My thoughts were the wrong way like "to simulate a `mouseenter` I have to dispatch a `mousesenter`"... I have figured out the events to dispatch now and rewritten all enter/leave tests.

I will look into the inner workings of the `ResponderEventPlugin` tomorrow to understand how I can trigger `getLowestCommonAncestor` with the values in the current test. The test renderer looks good, I'm sure I can get it to where I need it.

When I added the `react-native-renderer` to the test though I got the error `Invariant Violation: EventPluginRegistry: Cannot inject event plugin ordering more than once. You are likely trying to load more than one copy of React.` which makes sense. Is there any trick to _unload_ React or do I have to put the test in a different file?

gaearon commented:
>do I have to put the test in a different file?

This sounds right. In fact maybe you can move the original test for that part into `ResponderEventPlugin-test.internal.js` which uses internal APIs anyway. We can look at it later at some point.

timjacobi commented:
Sounds good. It looks like `ResponderEventPlugin-test.internal.js` needs some care anyway.
I have removed the WIP status now. Should be good to review.

gaearon commented on packages/events/__tests__/ResponderEventPlugin-test.internal.js:
```diff
@@ -1360,4 +1360,88 @@ describe('ResponderEventPlugin', () => {
     expect(ResponderEventPlugin._getResponder()).toBe(null);
   });
+
+  it('should determine the first common ancestor correctly', () => {
```
Since we're not using Responder plugin it's worth adding a comment explaining *why* the test is here and how it's related to Resonder.

gaearon commented on packages/react-dom/src/__tests__/ReactTreeTraversal-test.js:
```diff
@@ -0,0 +1,304 @@
+  describe('Two phase traversal', () => {
+    it('should not traverse when target is outside component boundary', () => {
+      const outerNode = document.createElement('div');
+      document.body.appendChild(outerNode);
```
The node is appended here to the document. We should remove it to clean up after the test. I usually put the rest of the test in try/finally, and remove it in the finally block.

gaearon commented on packages/react-dom/src/__tests__/ReactTreeTraversal-test.js:
```diff
@@ -0,0 +1,304 @@
+      var enterNode = document.createElement('div');
+
+      document.body.appendChild(leaveNode);
+      document.body.appendChild(enterNode);
```
Same.

gaearon replied:
In fact it may be easier to always add these nodes in beforeEach, and then always remove them in afterEach. Some tests will use them, some won't.

gaearon commented on packages/react-dom/src/__tests__/ReactTreeTraversal-test.js:
```diff
@@ -0,0 +1,304 @@
+      var leaveNode = document.createElement('div');
+      var enterNode = document.getElementById('P_P1_C1__DIV');
+
+      document.body.appendChild(leaveNode); // From the window or outside of the React sandbox.
```
If we change nodes to always be added, tests like this will have to *remove* one instead.

timjacobi replied:
Could you elaborate? I don't follow.

gaearon replied:
Never mind, I misread the tests.

timjacobi commented:
Updated according to comments :)

gaearon commented on packages/react-dom/src/__tests__/ReactTreeTraversal-test.js:
```diff
@@ -0,0 +1,293 @@
+  <div
+    id={id + '__DIV'}
+    onClickCapture={e =>
+      eventHandler(e.currentTarget.id, 'captured', e.type, ARG)
```
The logic behind choosing to put either `ARG` or `ARG2` here is not clear to me.

In the original tests, it was specific to a test. We're traversing enter/leave, and calling "from" path with `ARG` and "to" path with `ARG2`.

Here, we instead encoded `ARG` and `ARG2` right into components. I don't think this is right. This seems to couple them with specific test cases below. So it might be hard or impossible to add other completely valid tests which test the opposite enter/leave paths. This didn't use to be a problem.

I'm not sure what the solution is; just pointing out an issue I see.

timjacobi replied:
Yeah I'm not entirely happy with that either. Will think about a more elegant solution.

timjacobi replied:
I wonder if we're good to just remove the `ARG`s since we're now tracking the event type which we haven't done before. If I get it right`traverseEnterLeave(from, to, fn, argFrom, argTo)` will be called with `argFrom` and `argTo` set to the events that are dispatched on the target anyway so I feel like extracting the target id and event type will be strict enough.

gaearon replied:
Yes this sounds right

timjacobi commented:
@gaearon the App Veyor test failures seem unrelated.

gaearon commented:
Yeah, I didn't realize it would start appearing on PRs..

gaearon commented:
This is good. Thank you!

timjacobi commented:
No worries. Was great again!
